### PR TITLE
[ECO-4330] Feature - server initiated auth

### PR DESCRIPTION
**Pull request by sacOO7**
Fixed https://github.com/ably/ably-go/issues/228

## Summary by CodeRabbit

- **New Features**
	- Added a logger method to the Realtime struct for improved logging access.
	- Defined a new authentication action constant for better action management.

- **Improvements**
	- Enhanced error handling in the authorization process for more robust management of potential issues.
	- Updated connection state management logic for improved clarity and control during reauthorization and reconnection.

- **Tests**
	- Modified existing tests to align with new connection state expectations during reauthorization.
	- Added new tests to validate connection state transitions during reconnection scenarios.

### CODE REVIEW

File: ably/ably_test.go [hunk 1]
```diff
@@ -261,6 +261,24 @@ func (rec *MessageRecorder) CheckIfSent(action ably.ProtoAction, times int) func
...
+		counter := 0
+		for _, m := range rec.Received() {
+			if m.Action == action {
+				counter++
+				if counter == times {
+					return true
+				}
+			}
+		}
+		if times == 0 && times == counter {
```
Comment (amnonbc) [inline ably/ably_test.go:275]: what does this do?

File: ably/realtime_conn_spec_integration_test.go [hunk 2]
```diff
@@ -1684,18 +1694,214 @@ func TestRealtimeConn_RTN15h2_Success(t *testing.T) {
...
+	app, restClient := ablytest.NewREST()
+	defer safeclose(t, app)
+
+	recorder := NewMessageRecorder()
+	authCallbackTokens := []string{}
+
+	// Server sends AUTH message 30 seconds before token expiry.
+	// So sending client token with expiry of 33 seconds, server will send AUTH msg after 3 seconds.
+	authCallback := func(ctx context.Context, tp ably.TokenParams) (ably.Tokener, error) {
+		tokenExpiry := 33000
```
Comment (amnonbc) [inline ably/realtime_conn_spec_integration_test.go:1856]: I would make `tokenExpiry` a  `time.Duration`

Comment (sacOO7) [reply ably/realtime_conn_spec_integration_test.go:1856]: Ttl accepts int value in terms of milliseconds. Time.duration is measured in nanoseconds

File: ably/ably_test.go (hunk 1 above)
Comment (sacOO7) [reply ably/ably_test.go:275]: If 0 is passed as times integer, that means it expects given proto msg type to be called 0 times.

File: ably/realtime_conn_spec_integration_test.go (hunk 2 above)
Comment (amnonbc) [reply ably/realtime_conn_spec_integration_test.go:1856]: normally one one do something like
```
tokenExpery := 33 * time.Second
token, err := restClient.Auth.RequestToken(context.Background(), &ably.TokenParams{TTL: tokenExpiry.Milliseconds()})
```

Comment (sacOO7) [reply ably/realtime_conn_spec_integration_test.go:1856]: Ohkay, let me check on this one

File: ably/ably_test.go (hunk 1 above)
Comment (sacOO7) [reply ably/ably_test.go:275]: Also means, we are not expecting protoMsg with given action to be received.

File: ably/realtime_conn_spec_integration_test.go (hunk 2 above)
Comment (sacOO7) [reply ably/realtime_conn_spec_integration_test.go]: updatedhttps://github.com/ably/ably-go/pull/661/commits/edce32f870634997fffa86c28687e644baecc8b3

File: ably/ably_test.go (hunk 1 above)
Comment (amnonbc) [reply ably/ably_test.go:275]: would it be clearer as 
`if times == 0 && counter == 0 {`
?

**Review by amnonbc — APPROVED** [review]
nice work.

File: ably/ably_test.go (hunk 1 above)
Comment (sacOO7) [reply ably/ably_test.go]: updated https://github.com/ably/ably-go/pull/661/commits/bd39bc370450449fd2f6140d721428ca69173b5b

**Pull request merged**